Added doc describing versioning and release process (#10)
strimzi/governance: RELEASES.md

# Releases

This document describes how Strimzi handles versioning, releases new versions and communicates about them.

> Strimzi uses [semantic versioning](https://semver.org/) as a versioning schema for the releases of all the components.

Reviewer: ```suggestion
All Strimzi component releases use [semantic versioning](https://semver.org/).
```

## Release Cycle

> Strimzi is using a release schedule of 1/2 months for the "core" component, the cluster operator.

Reviewer 2: `1/2`? Should it be `1 or 2`?

> Strimzi is using a release schedule of 1/2 months for the "core" component, the cluster operator.

Reviewer: ```suggestion
Strimzi is uses a 1–2 month release schedule for the "core" operators component.
```

The [GitHub milestones](https://github.com/strimzi/strimzi-kafka-operator/milestones) are used to track the features and bug fixes included in each release.

> The other components, under the Strimzi umbrella, don't have a specific release cycle but when having a new releases is decided on a case-by-case basis depending on the new features added, bugs and vulnerabilities fixed.

Reviewer: ```suggestion
Other Strimzi components don't have a specific release cycle. 
New releases are decided on a case-by-case basis depending on the new features added, bugs and vulnerabilities fixed.
```

## Release Process

> Every Strimzi component has a corresponding documentation about how to make a new release.

Reviewer 2: I do not think we have this in every repository.

For example, the release process for the cluster operator is described [here](https://github.com/strimzi/strimzi-kafka-operator/blob/main/development-docs/RELEASE.md).

## Keeping Track Of Changes

> We provide a changelog that includes all changes for each version taking the history from the first release.

Reviewer 2: It does not include all changes ... only the things which we saw a notable.

It is available as part of the repository of each component, for example, here is the changelog for the [Strimzi cluster operator](https://github.com/strimzi/strimzi-kafka-operator/blob/main/CHANGELOG.md).

Every closed issue and PR should be added to a GitHub Milestone that provides an overview of everything that will be included in the corresponding release.

Lastly, every new release will come with a GitHub Release which provides the exact changelog for that release.

## Get Notified For New Releases

Every new release, for each component, is announced in several ways:

* Email to the Strimzi users CNCF mailing list.
* Strimzi CNCF Slack channel.

> * Twitter [account](https://twitter.com/strimziio).

Reviewer 2: We don't do this really. I usually tweet about it because I did the video for new minor releases. But I don't think we really tweet for every release.

Author: I was pretty sure about it, but it seems I was wrong. I will remove this bullet point then.

Another way to be notified is by [watching](https://docs.github.com/en/github/managing-subscriptions-and-notifications-on-github/setting-up-notifications/configuring-notifications#configuring-your-watch-settings-for-an-individual-repository) our Strimzi repositories on GitHub, for example the [Strimzi cluster operator](https://github.com/strimzi/strimzi-kafka-operator) one.

## Removing Features, Breaking Changes, Deprecations and Feature Gates

> Every new major version, Strimzi maintainers are allowed to remove features or make breaking changes as per [semantic versioning](https://semver.org/).

Reviewer 2: I'm not sure I would refer to this too much here. The SemVer actually says that `0.y.z` versions do not define or make any API promises. So by suggesting we follow SemVer, it might be understood as _at this point there are no guarantees_. So I would think just describing the process would be better.

Author: Makes sense. I will just remove the above sentence then.

When there is an API deprecation, because it is going to be replaced by a new one, it is announced in the coming release changelog. 
We define a plan about when that API will be completely removed in the future giving the users enough time and the steps needed to adopt the new one.

New features could be added through the so called "feature gates".

> It means that the feature is disabled by default but users can start trying it by enabling the corresponding gate.

Reviewer 2: ```suggestion
It means that the feature is disabled by default at the beginning but users can start trying it by enabling the corresponding gate.
```

> Even in this case, we describe the plan about when that feature will be enabled by default and removed from the "feature gate".

Reviewer: This is a bit specific to the operator (e.g. talking about feature gates). Really for each component we should define:

* What we consider to be the "public API" that being covered by semver. For the operator that's the CR schemas + the operator configuration "schemas" + the config to do with non-alpha feature gates. But for oauth it's the configuration "schema". Etc.
* How much how many minor releases of notice should be required between deprecation and removal. For instance, is it OK to deprecate in `x.y.0` and remove `x.(y+1).0`?
I guess we could define this in a file per repo if we wanted, so avoid `governance` having to know about each project.

Author: You are right, feature gates are specific to CO and we could re-arrange the sentence above to make it clear.

Regarding the API, isn't it enough to add something like:

"By API it would mean the Custom Resources schemas and configuration for "core" operators, OpenAPI definition for the HTTP bridge or configuration for the OAuth component".

Regarding the versions related to deprecation and remove I think it's usually described in the "proposal", not sure we have a strict rule.

Reviewer: > "By API it would mean the Custom Resources schemas and configuration for "core" operators, OpenAPI definition for the HTTP bridge or configuration for the OAuth component".

OK, let's add wording like this for now. 

> Regarding the versions related to deprecation and remove I think it's usually described in the "proposal", not sure we have a strict rule.

Then can we say something like "The notice period between deprecation and removal is defined on a case-by-case basis as part of the proposal or PR that deprecates an API".

# Releases

This document describes how Strimzi handles versioning, releases new versions and communicates about them.

Strimzi uses [semantic versioning](https://semver.org/) as a versioning schema for the releases of all the components.

## Release Cycle

Strimzi is using a release schedule of 1 or 2 months for the "core" component, the cluster operator.

> The [GitHub milestones](https://github.com/strimzi/strimzi-kafka-operator/milestones) are used to track the features and bug fixes included in each release.

Reviewer 2: Do we want to also point to our pathetic roadmap? https://github.com/orgs/strimzi/projects/1

Author: I never know how much it's updated ... shame on me but I don't take a look often to it. Interesting you said it's pathetic :-D ... anyway if you want I can mention it.

Reviewer 2: Well, the roadmap sucks ... and to be honest our planning is bad. But I think anyone seeing this will ask for it. So I think we need to do something with it.

Reviewer: Would it help the planning and keeping it updated if it were part of the release process to do so?

> The other components, under the Strimzi umbrella, don't have a specific release cycle but when having a new releases is decided on a case-by-case basis depending on the new features added, bugs and vulnerabilities fixed.

Reviewer 2: _don't have a specific release cycle_ sounds weird to me. Would maybe this sound a bit nicer?

```suggestion
The other components, under the Strimzi umbrella, are released on a case-by-case basis depending on the new features added, bugs and vulnerabilities fixed.
```

Author: Agree, I didn't like it neither :-)

## Release Process

Most of the Strimzi components have a corresponding documentation about how to make a new release.

> For example, the release process for the cluster operator is described [here](https://github.com/strimzi/strimzi-kafka-operator/blob/main/development-docs/RELEASE.md).

Reviewer 2: Should we talk here about release candidates? Where they are announced, when are they done, how long at minimum will be given for testing etc.?

Author: Taking a look at other projects, i.e. KEDA, I didn't see RCs mentioned there but I have no strong opinion about not mentioning them. The announcement follow the same process and I guess 2-3 days are enough for testing?

Author: I have written something, let's see if that's ok.

# Releases

This document describes how Strimzi handles versioning, releases new versions and communicates about them.

Strimzi uses [semantic versioning](https://semver.org/) as a versioning schema for the releases of all the components.

## Release Cycle

Strimzi is using a release schedule of 1 or 2 months for the "core" component, the cluster operator.
The [GitHub milestones](https://github.com/strimzi/strimzi-kafka-operator/milestones) are used to track the features and bug fixes included in each release.
There is also a [roadmap](https://github.com/orgs/strimzi/projects/1) to describe the features that will be available in the future releases.

The other components, under the Strimzi umbrella, are released on a case-by-case basis depending on the new features added, bugs and vulnerabilities fixed.

## Release Process

> Most of the Strimzi components have a corresponding documentation about how to make a new release.

Reviewer: ```suggestion
Most of the Strimzi components have documentation about how to make a new release.
```

Can we make this `RELEASE.md` for each component? It would make this document seem less _laissez faire_.

Author: Why? I think that the release process is quite the same across components and this doc seems to cover it. Each repo has its own doc for actually "doing" the release.

Reviewer: What I meant was, can we make `RELEASE.md` the standard for all our projects, so that this sentence can read 
```suggestion
All Strimzi components have a documentation for their release process in their `RELEASE.md` file.
```

Author: Currently each component has a document like this ... i.e. bridge
https://github.com/strimzi/strimzi-kafka-bridge/blob/main/development-docs/RELEASING.md
It explains the steps for making the release. While you want this RELEASE.md duplicated across all repos with the difference in the process for each one? :-( ... not sure we really need it tbh.

Reviewer: I think one of the two of us is confused. It might be me. But I think if every project already has a `RELEASING.md` file, then all I'm asking for here is that we say that here. The problem with "Most of the Strimzi components have documentation..." is it raises the question "Why don't all of them". And the current sentence doesn't make it easier for a reader to know where this documentation lives. So I'm just asking that we state what the file name is.

Author: Ok I was making the point about the difference between the current `RELEASES.md` and the `RELEASING.md` file we have in the other repository.
The `RELEASING.md` describe the actual steps to do a build (i.e. branch creation, tagging, Azure pipeline runs and so on).
The `RELEASES.md` is about having an overall description of how releases and versioning work in Strimzi.
I thought you were referring to have a `RELEASES.md` for each repository.

Anyway I just double checked. Operator, bridge and canary having a `RELEASING.md`. OAuth calls is `RELEASE.md`. Other repos (i.e. quota plugin) don't have something like that.

Reviewer: Let's change the Oauth `RELEASE.md` to `RELEASING.md` for consistency (and because `RELEASE.md` is confusingly close to `RELEASES.md`, which we've just established has a different purpose).

Author: Agree. Just opened https://github.com/strimzi/strimzi-kafka-oauth/pull/139

For example, the release process for the cluster operator is described [here](https://github.com/strimzi/strimzi-kafka-operator/blob/main/development-docs/RELEASE.md).

Before an official release, at least one Release Candidate (RC) is built for testing.
More RCs can follow the first one if new bugs are found along the way.
The Release Candidates are announced on the Slack channel and Strimzi user CNCF mailing list, so that the community can test before the final release.

> The maintainers provides 2-3 days to the community for testing and, if no objections are raised, the final release is announced.

Reviewer: Comparing this to the [Kafka release process](https://cwiki.apache.org/confluence/display/KAFKA/Release+Process) (see also [the Kafka bylaws](https://cwiki.apache.org/confluence/display/KAFKA/Bylaws)), it's not so clear _who_ gets to decide about the validity of any objections. In fact, taken literally, any random person/bot on the internet could "object" and prevent a release. I'm not sure we want/need to document every last detail, but it would be good to be _bit_ clearer. For example, making the actual final release dependent on a maintainer vote makes it clearer who decides about any objections, and I don't think it would slow down releases very much, because we already seek a consensus on releases, just not publicly enough.

Author: So are you suggesting to add a sentence like ...

"If any objections, maybe related to a bug, there will be a vote from maintainers to decide to stop or move forward with the release". or "If any objections, maybe related to a bug, the maintainers have to reach a consensus on stopping or moving forward with the release".

Or do you want something more specific about how it works, via email, slack channel ... ?

Reviewer 2: Vote suggests a formal process which you would need to define. We should either really setup a proper release voting process. So I guess you can say something like _any objections will be evaluated and decided on_?

Not sure anyone would raise it in the incubation - but of course Jakub Scholz announcing RC and without any sort of feedback announcing GA is not exactly a process. A process following Apache style voting as for example used in Kafka would be obviously more advanced. But the question is whether we have enough resources to change it.

Reviewer: What resources does it need other than maintainers paying attention?

Reviewer 2: > What resources does it need other than maintainers paying attention?

The maintainers and their time are the resources here.

It is available as part of the repository of each component, for example, here is the changelog for the [Strimzi cluster operator](https://github.com/strimzi/strimzi-kafka-operator/blob/main/CHANGELOG.md).

Every closed issue and PR should be added to a GitHub Milestone that provides an overview of everything that will be included in the corresponding release.

Lastly, every new release will come with a GitHub Release which provides the exact changelog for that release.

## Get Notified For New Releases

Every new release, for each component, is announced in several ways:

* Email to the Strimzi users CNCF mailing list.
* Strimzi CNCF Slack channel.

Another way to be notified is by [watching](https://docs.github.com/en/github/managing-subscriptions-and-notifications-on-github/setting-up-notifications/configuring-notifications#configuring-your-watch-settings-for-an-individual-repository) our Strimzi repositories on GitHub, for example the [Strimzi cluster operator](https://github.com/strimzi/strimzi-kafka-operator) one.

## Removing Features, Breaking Changes, Deprecations and Feature Gates

> When there is an API deprecation, because it is going to be replaced by a new one, it is announced in the coming release changelog. 

Reviewer: We shouldn't limit this to APIs which _get_ a replacement. Let's allow ourselves to remove API/features when we decide we messed up.
```suggestion
When there is an API deprecation (with out without a replacement) it is announced in the coming release changelog. 
```

We define a plan about when that API will be completely removed in the future giving the users enough time and the steps needed to adopt the new one.

> New features could be added through the so called "feature gates".

Reviewer: ```suggestion
New features may be added through [feature gates](https://github.com/strimzi/proposals/blob/main/022-feature-gates.md).
```

It means that the feature is disabled by default at the beginning but users can start trying it by enabling the corresponding gate.
Even in this case, we describe the plan about when that feature will be enabled by default and removed from the "feature gate".

## Apache Kafka Support

On each new Strimzi cluster operator release, it is not feasible to maintain support for all the available Apache Kafka versions.

> There is a well defined plan for supported Apache Kafka versions you can read [here](https://github.com/strimzi/strimzi-kafka-operator/blob/main/KAFKA_VERSION_SUPPORT.md).

Reviewer: ```suggestion
There is a well-defined plan for supported Apache Kafka versions you can read [here](https://github.com/strimzi/strimzi-kafka-operator/blob/main/KAFKA_VERSION_SUPPORT.md).
```
